tardis/simulation/base.py

```diff
@@ -401,6 +406,137 @@ def to_hdf(self, model, path_or_buf, path='', plasma_properties=None):
         self.runner.to_hdf(path_or_buf, path)
         model.to_hdf(path_or_buf, path, plasma_properties)
 
+    def make_source_function(self, model):
+        """
+        Calculates the source function using the line absorption rate estimator `Edotlu_estimator`
+
+        Formally it calculates the expresion ( 1 - exp(-tau_ul) ) S_ul but this product is what we need later,
+        so there is no need to factor out the source function explicitly.
+
+        Parameters
+        ----------
+        model : tardis.model.Radial1DModel
+
+        Returns
+        -------
+        DataFrame containing ( 1 - exp(-tau_ul) ) S_ul
+        """
+
+        upper_level_index = model.atom_data.lines.set_index(['atomic_number', 'ion_number', 'level_number_upper']).index.copy()
+        e_dot_lu = pd.DataFrame(self.runner.Edotlu, index=upper_level_index)
+        e_dot_u = e_dot_lu.groupby(level=[0, 1, 2]).sum()
+        e_dot_u.index.names = ['atomic_number', 'ion_number', 'source_level_number'] # To make the q_ul e_dot_u product work, could be cleaner
+        transitions = model.atom_data.macro_atom_data[model.atom_data.macro_atom_data.transition_type == -1].copy()
+        transitions_index = transitions.set_index(['atomic_number', 'ion_number', 'source_level_number']).index.copy()
+        tmp = model.plasma.transition_probabilities[(model.atom_data.macro_atom_data.transition_type == -1).values]
+        q_ul = tmp.set_index(transitions_index)
+        return (model.atom_data.lines.wavelength[transitions.transition_line_id].values.reshape(-1,1) * 
+                 (q_ul * e_dot_u) * model.tardis_config.supernova.time_explosion / (4*np.pi) )
+
+    def integrate(self,model):
+        num_shell, = self.runner.volume.shape
+        ps         = np.linspace(0.999, 0, num = 3 * num_shell)
+        R_max      = self.runner.r_outer_cgs.max()
+        R_min_rel  = self.runner.r_inner_cgs.min() / R_max
+        ct         = co.c.cgs.value * self.tardis_config.supernova.time_explosion.value / R_max
+        J_blues    = model.j_blues
+        J_rlues    = model.j_blues.values * np.exp( -model.plasma.tau_sobolevs.values) + self.runner.att_S_ul
+
+        r_shells = np.zeros((num_shell+1,1))
+        # Note the reorder from outer to inner
+        r_shells[1:,0],r_shells[0,0] = self.runner.r_inner_cgs[::-1] / R_max, 1.0
+        z_crossings = np.sqrt(r_shells**2 - ps**2)
+
+        z_ct = z_crossings/ct
+        z_ct[np.isnan(z_crossings)] = 0
```

Comment:
I see, r_shells is a 2D array. Do you need it to be a 2D array?
Because in 447 you are only setting one dimension.

Please enlighten me on what `z_crossings` should look like? What dimension do you expect?